[CASSANDRA-20052][4.1] Introduce CQL message limits in V5 protocol logic

netudima opened the pull request:
Add native_transport_max_message_size and native_transport_max_auth_message_size configuration options to limit maximum CQL message size for V5-protocol ore later. 
Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20052

netudima committed:
Introduce CQL message limits in V5 protocol logic

Add native_transport_max_message_size and native_transport_max_auth_message_size configuration options to limit maximum CQL message size for V5-protocol ore later.
Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20052

Commit:
Skip too big non-auth messages without connection closing, return pausing of consumption for big messages in rate limiting logic, validate correspondent configuration parameters alignment

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe for CASSANDRA-20052

beobal commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -269,6 +269,8 @@ public MemtableOptions()
     @Replaces(oldName = "native_transport_max_frame_size_in_mb", converter = Converters.MEBIBYTES_DATA_STORAGE_INT, deprecated = true)
     public DataStorageSpec.IntMebibytesBound native_transport_max_frame_size = new DataStorageSpec.IntMebibytesBound("16MiB");
+    public volatile DataStorageSpec.LongBytesBound native_transport_max_message_size = null;
+    public volatile DataStorageSpec.LongBytesBound native_transport_max_auth_message_size = new DataStorageSpec.LongBytesBound("128KiB");
```
Rather than having two separate settings for max message and max auth message size, why don't we enforce a rule that any message received before the server is ready must fit in a single frame?
That would naturally limit those messages to 128KiB and would not need operators to have to think about what is a reasonable max size for auth messages. If we really want to accommodate an authentication scheme which does require large messages, we can add a boolean setting to allow large messages, which I think would still be conceptually simpler, in which case auth messages would still be subject to the `MAX_CQL_MESSAGE_SIZE` check.

netudima replied:
yes, I think it is a good idea, I will apply it

beobal commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -182,6 +190,13 @@ protected boolean processOneContainedMessage(ShareableBytes bytes, Limit endpoin
 
         // max CQL message size defaults to 256mb, so should be safe to downcast
         int messageSize = Ints.checkedCast(header.bodySizeInBytes);
+        if (authMessageTooBig(messageSize))
```
Any message being handled here *must* be < 128KiB or else it would not fit in a single frame, so I would argue we don't need this check at all. Do you think there's a use case where operators might set the max size of an auth message smaller than this?

netudima replied:
yes, I though about cases when a lot of parallel requests to consume a lot of memory by authentication requests are issued as a kind of DoS attack + usually the normal auth message size is not more than few KiB but probably I am overprotective here :-) and having a limit to 1 frame for auth  messages is enough.

beobal commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -518,10 +533,23 @@ protected boolean processFirstFrameOfLargeMessage(IntactFrame frame, Limit endpo
             int messageSize = Ints.checkedCast(header.bodySizeInBytes);
             receivedBytes += buf.remaining();
+
+            if (authMessageTooBig(messageSize))
```
I would prefer we handle this by disallowing "auth" messages extending beyond a single frame. That means strictly that `STARTUP, CREDENTIALS, OPTIONS, AUTH_RESPONSE` message received here can be considered too large and rejected, regardless of their actual size. This would also have the advantage that we can identify the message type by inspecting the `Header` and we would no longer need to hold a reference to the `ServerConnection`.

netudima replied:
Here we may have a security issue:  the connection state machine which checks what request types are valid currently is within ServerConnection, so a malicious user can send a request with a type != STARTUP, CREDENTIALS, OPTIONS, AUTH_RESPONSE (for example BATCH) to bypass this size protection, in other words we cannot rely on an input from network such as request type to decide if a size check is needed.

beobal replied:
Excellent point +1 to keeping it based on `ServerConnection.stage()`

beobal commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -518,10 +533,23 @@ protected boolean processFirstFrameOfLargeMessage(IntactFrame frame, Limit endpo
             LargeMessage largeMessage = new LargeMessage(header);
-
-            if (throwOnOverload)
+            if (messageSize > MAX_CQL_MESSAGE_SIZE)
```
This is a good defence against accepting a mutation bigger than the max size that the commitlog will accept, but I think we can go further to protect against runaway resource consumption. 

I had imagined that we would override `onIntactFrame` and the `tooBig` flag would have the effect that the `LargeMessage` would not hold on to the individual frame buffers, knowing that it was not going process the message (it could also do the same if the message is marked overloaded).

netudima replied:
totally agree, actually I have come to the same conclusion as well and I am already working now on such change

netudima replied:
there is one more thing to fix actually - handleErrorAndRelease invocation is not correct in my case onComplete, because there is no acquired capacity to release it back, so it should be handleError. Probably it is applicable for overload == Overload.BYTES_IN_FLIGHT as well but I need to double check

beobal commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -822,6 +822,25 @@ else if (conf.commitlog_segment_size.toMebibytes() >= 2048)
+            Math.min(conf.max_mutation_size.toBytes(),
+                     Math.min(
+                        conf.native_transport_max_request_data_in_flight.toBytes(),
+                        conf.native_transport_max_request_data_in_flight_per_ip.toBytes()
```
These two settings can be modified at runtime, but `native_transport_max_messages_size` can not. Even if it could, it's memoized in `CQLMessageHandler.MAX_CQL_MESSAGE_SIZE` which is final.

netudima committed:
Address review comments: remove native_transport_max_auth_message_size, update nativeTransportMaxMessageSizeInBytes if rate limiting parameters are changed in runtime, do not store buffers for a message in case of it is too big or it is an overload error, allow only single-frame messages during an authentication stage

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe for CASSANDRA-20052

beobal committed:
Minor simplification

beobal commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -747,18 +773,85 @@ private void markBackpressure(Overload backpressure)
+        }
+    }
+
+    private boolean isItLargeMessageSentDuringAuth(Envelope.Header header)
```
I pushed a commit here: https://github.com/beobal/cassandra/commit/73b8895ae6c3312165330198cc4d9e720c471c98 which inlines this, seeing as it's now only used in a single place.

netudima replied:
thanks, looks good to me, should I add the changes to the current MR?

beobal replied:
sure, if you're happy with that.

netudima replied:
I cherry-picked the change and pushed it into the MR

beobal reviewed (commented):
This looks good to me, please see what you think about the proposed minor change I mentioned.

netudima committed:
Fix failing unit tests: adjust the logic to test overload by emulating concurrent processing of more than one message

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe for CASSANDRA-20052

beobal commented on test/unit/org/apache/cassandra/transport/ClientResourceLimitsTest.java:
```diff
@@ -261,26 +284,39 @@ public void testOverloadedExceptionWhenGlobalLimitByMultiFrameMessage()
         // Make sure we can only exceed the per-endpoint limit
         ClientResourceLimits.setGlobalLimit(HIGH_LIMIT);
-        testOverloadedException(() -> client(true, Ints.checkedCast(LOW_LIMIT / 2)));
+        // test massage = 2/3 x
```
nit: typo

beobal commented on test/unit/org/apache/cassandra/transport/ClientResourceLimitsTest.java:
```diff
@@ -261,26 +284,39 @@ public void testOverloadedExceptionWhenGlobalLimitByMultiFrameMessage()
         // Bump the per-endpoint limit to make sure we exhaust the global
         ClientResourceLimits.setEndpointLimit(HIGH_LIMIT);
-        testOverloadedException(() -> client(true, Ints.checkedCast(LOW_LIMIT / 2)));
+        // test massage = 2/3 x
```
nit: typo

beobal commented on test/unit/org/apache/cassandra/transport/RateLimitingTest.java:
```diff
@@ -155,7 +171,11 @@ private void testBytesInFlightOverload(int payloadSize) throws Exception
             StorageService.instance.setNativeTransportRateLimitingEnabled(true);
             ClientResourceLimits.GLOBAL_REQUEST_LIMITER.setRate(OVERLOAD_PERMITS_PER_SECOND, ticker);
-            ClientResourceLimits.setGlobalLimit(1);
+            // test massage = 1x
```
nit: typo

beobal commented on test/unit/org/apache/cassandra/transport/RateLimitingTest.java:
```diff
@@ -155,7 +171,11 @@ private void testBytesInFlightOverload(int payloadSize) throws Exception
-            ClientResourceLimits.setGlobalLimit(1);
+            // test massage = 1x
+            // emulated concurrent message = 1.5x
+            // test massage + emulated concurrent message = 2.5x > 2x set as a global limit
```
nit: typo

beobal reviewed (approved):
lgtm (there is a typo, replicated in several comments)

netudima committed:
Fix comment typos

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe for CASSANDRA-20052

netudima commented:
thanks, the typos are fixed

maedhroz commented on test/unit/org/apache/cassandra/transport/AuthMessageSizeLimitTest.java:
```diff
@@ -0,0 +1,178 @@
+        StringBuilder password = new StringBuilder(length);
+        for (int i = 0; i < length; i++)
+            password.append('a');
+        return password.toString();
```
Fun thing I just learned: You can replace all this with:

```
return "a".repeat(length);
```

(Appears above and in `MessageSizeLimitTest` too.)

netudima replied:
yes, it is a useful trick, unfortunately the method has been added only in Java 11 (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#repeat(int)), so I cannot use it for Cassandra 4.x

maedhroz commented on test/unit/org/apache/cassandra/transport/AuthMessageSizeLimitTest.java:
```diff
@@ -0,0 +1,178 @@
+        return new QueryMessage(query.toString(), V5_DEFAULT_OPTIONS);
+    }
+
+    private static String createIncorrectLongPassword(int length) {
```
nit: `{` on new line

netudima replied:
fixed

maedhroz commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -822,6 +825,21 @@ else if (conf.commitlog_segment_size.toMebibytes() >= 2048)
+            nativeTransportMaxMessageSizeConfiguredExplicitly = true;
+            long maxCqlMessageSize = conf.native_transport_max_message_size.toBytes();
+            if (maxCqlMessageSize > conf.native_transport_max_request_data_in_flight.toBytes())
+                throw new ConfigurationException("native_transport_max_message_size must no exceed native_transport_max_request_data_in_flight", false);
```
```suggestion
                throw new ConfigurationException("native_transport_max_message_size must not exceed native_transport_max_request_data_in_flight", false);
```

netudima replied:
fixed

maedhroz commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -822,6 +825,21 @@ else if (conf.commitlog_segment_size.toMebibytes() >= 2048)
+                throw new ConfigurationException("native_transport_max_message_size must no exceed native_transport_max_request_data_in_flight", false);
+
+            if (maxCqlMessageSize > conf.native_transport_max_request_data_in_flight_per_ip.toBytes())
+                throw new ConfigurationException("native_transport_max_message_size must no exceed native_transport_max_request_data_in_flight_per_ip", false);
```
```suggestion
                throw new ConfigurationException("native_transport_max_message_size must not exceed native_transport_max_request_data_in_flight_per_ip", false);
```

netudima replied:
fixed

maedhroz commented on test/unit/org/apache/cassandra/transport/ClientResourceLimitsTest.java:
```diff
@@ -137,6 +143,23 @@ private SimpleClient client(boolean throwOnOverload, int largeMessageThreshold)
+        ClientResourceLimits.ResourceProvider resourceProvider = new ClientResourceLimits.ResourceProvider.Default(allocator);
+        resourceProvider.globalLimit().release(length);
+        resourceProvider.endpointLimit().release(length);
+    }
```
nit: Noting that this is duplicated in `RateLimitingTest`. Might be something we could move to a shared abstract class, along with query options and client creation, but not a big deal if you don't think it's worth it. `MessageSizeLimitTest` and `AuthMessageSizeLimitTest` share some logic as well. Perhaps `abstract MessageLimitTest extends CQLTester` or `abstract NativeProtocolLimitTest extends CQLTester` for these 4?

netudima replied:
agree, I have extracted some common logic from these 4 classes

maedhroz commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -747,18 +779,71 @@ private void markBackpressure(Overload backpressure)
+            this.tooBig = true;
+        }
+
+        protected void onIntactFrame(IntactFrame frame)
```
nit: `@Override`

maedhroz commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -747,18 +779,71 @@ private void markBackpressure(Overload backpressure)
+                super.onIntactFrame(frame);
+        }
+
+        protected void onCorruptFrame()
```
nit: `@Override`

maedhroz commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -747,18 +779,71 @@ private void markBackpressure(Overload backpressure)
+        }
+
+
         protected void onComplete()
```
nit: `@Override`

maedhroz commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -747,18 +779,71 @@ private void markBackpressure(Overload backpressure)
                 processRequest(assembleFrame(), backpressure);
         }
 
         protected void abort()
```
nit: `@Override`

maedhroz commented on src/java/org/apache/cassandra/transport/CQLMessageHandler.java:
```diff
@@ -518,10 +523,29 @@ protected boolean processFirstFrameOfLargeMessage(IntactFrame frame, Limit endpo
+                // not make sense to continue processing subsequent frames
+                handleError(ProtocolException.toFatalException(new OversizedAuthMessageException(
+                            "The connection is not yet in a valid state to process multi frame CQL Messages, usually this" +
+                            "means that authentication is still pending. " +
```
nit: It might be useful to create a constant string from this sentence, then check that the exception message contains it in `AuthMessageSizeLimitTest` w/ `Assertions.assertThatThrownBy().isInstanceOf().hasMessageContaining()`. I suppose there might be other `ProtocolException` possibilities, and the test could pass by mistake...

maedhroz reviewed (approved):
Just a few nits, but LGTM

Commit:
address review comments: fix typos in error messages, fix formatting issues, a minor refactoring of tests related to CQL rate limiting

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe, Caleb Rackliffe for CASSANDRA-20052

netudima commented:
@maedhroz  thank you a lot for the review! I have committed changes to address your comments

Commit:
mark the extracted common test class as abstract

Patch by Dmitry Konstantinov; reviewed by Sam Tunnicliffe, Caleb Rackliffe for CASSANDRA-20052
